### ISSUE TITLE
out of space building compressed backup: need option to control location of temp.zip

### ISSUE THREAD

**Issue by sandsify** [robweber/xbmcbackup#83]
IOError: (28, 'No space left on device')         

A common use case is that you want to backup to a mounted HDD. This is typically the case when there is limited space.  But if you select option for a compressed archive the script attempts to build the zip xbmc_backup_temp.zip on local storage ~/.kodi/temp which causes it to run out of space. In this case we need to be able to host the temporary files on the external HDD as well.

 IOError: (28, 'No space left on device')  

openelec 6.0.3/ kodi 15 on wetek device

**Comment by robweber** [comment]
Also referenced is #78

**Comment by W4CROBSTER** [comment]
Whats the fastest way to restore a built?  computer or dropbox or ...

**Comment by robweber** [comment]
@W4CROBSTER - not sure what you mean by this, it's not really related to the issue at all. Please use the Kodi forum for general questions: http://forum.kodi.tv/showthread.php?tid=129499

**Comment by Cris--** [comment]
I think i have the same issue on my RPi3+LibreElec 8.1.2.
Can't check the compress option. Every try with it gives error and stops the backup process!
Doesn't matter where the backup location is (at the moment a cifs folder is via system.d mounted).

`EXCEPTION Thrown (PythonToCppException) : -->Python callback/script returned the following error<--
 - NOTE: IGNORING THIS CAN LEAD TO MEMORY LEAKS!
    Error Type: <type 'exceptions.MemoryError'>
    Traceback (most recent call last):
       File "/storage/.kodi/addons/script.xbmcbackup/scheduler.py", line 156, in <module>
       BackupScheduler().start()
       File "/storage/.kodi/addons/script.xbmcbackup/scheduler.py", line 57, in start
       backup.run(XbmcBackup.Backup,True)
       File "/storage/.kodi/addons/script.xbmcbackup/resources/lib/backup.py", line 256, in run
       filesCopied = self.backupFiles(fileGroup['files'],self.xbmc_vfs,self.remote_vfs)
       File "/storage/.kodi/addons/script.xbmcbackup/resources/lib/backup.py", line 482, in backupFiles
       wroteFile = dest.put(aFile,dest.root_path + aFile[len(source.root_path):])
       File "/storage/.kodi/addons/script.xbmcbackup/resources/lib/vfs.py", line 97, in put
       self.zip.writestr(utils.encode(dest),aFile.read())
       File "/usr/lib/python2.7/zipfile.py", line 1243, in writestr
    MemoryError
-->End of Python script error report<--`

**Comment by Gaet81** [comment]
Hi,

This would be a very handy feature for people runing on a RPI.
Is there any progress on this?

Thanks

**Comment by robweber** [comment]
This is a much requested feature. The issue I have with implementation are issues with the zip implementation python. 

In order to create the zip file it has to be staged somewhere. Ideally I'd like to write the zip directly to the destination on the fly, however the xbmcvfs module won't do this. It only handles basic file processes like copy, remove, list, etc. So when using the python zip module I have to give it a path to stage the zip file before the copy to the remote repository. The zip module will only write to a local type of file system (or a system that appears local). 

I can offer the ability to place the zip file in a different location, however the limitation would be that it has to be local. **IF** you can use a mount command or windows network share to make an SMB drive appear local this would work. Would this even be worth it for most people? I'm really not sure. Thoughts?

**Comment by Morphy99** [comment]
> This is a much requested feature. The issue I have with implementation are issues with the zip implementation python.
> 
> In order to create the zip file it has to be staged somewhere. Ideally I'd like to write the zip directly to the destination on the fly, however the xbmcvfs module won't do this. It only handles basic file processes like copy, remove, list, etc. So when using the python zip module I have to give it a path to stage the zip file before the copy to the remote repository. The zip module will only write to a local type of file system (or a system that appears local).
> 
> I can offer the ability to place the zip file in a different location, however the limitation would be that it has to be local. **IF** you can use a mount command or windows network share to make an SMB drive appear local this would work. Would this even be worth it for most people? I'm really not sure. Thoughts?

I'd like to be able to change this as well. If we can specify a location we could use a symlinked dir to an external drive .

**Comment by robweber** [comment]
I get the frustration but I really struggle with this from a usability standpoint. This is a really advanced request that could easily not work for people that don't understand the nuances of local vs remote folders on their systems. I'm currently re-working the settings system to use the basic/standard/advanced/expert modes much like the other Kodi settings areas (#176). Once that is done I think this might be worth the effort as I can include some help text and hide it within the Expert editor.

**Comment by Morphy99** [comment]
Is changing the temp location that advanced? USB drives are auto-mounted in Libreelec so it would be much like selecting which folders to backup?

**Comment by robweber** [comment]
The implementation is trivial, the usage is advanced. I always hesitate to add something that 99% of users will mess up trying to use properly. I think the settings levels will really help in this case though.

**Comment by Morphy99** [comment]
Sorry, I think I'm missing something, how would changing the temp location cause any issues for anyone?

**Comment by robweber** [comment]
It's sort of summarized above. It HAS to be a local directory or the built in python zip file system won't work. I can see lots of people choosing non-local systems (smb folders for example) and having issues with it not working and throwing errors.

**Comment by Morphy99** [comment]
Ah yes OK I guess Kodi does make accessing remote file systems easy. I can't think that many would want to change it but if you feel like it's best hiding, your call.

**Comment by robweber** [comment]
Merged this into the matrix branch only. Info on it's use also [added to the wiki](https://github.com/robweber/xbmcbackup/wiki/Configuration). Main points are: 

1. It must be a local directory to the device you are running Kodi on. A mounted path like ```H:\foo``` or ```/mnt/bar``` should probably work. Use of Kodi paths like ```smb://host/path``` will fail. This is a limitation of the Python zip library. 
2. Don't put a path that you're also backing up. This will create a recursion. 

This is an expert setting, use at your own risk.
